File: hololive/hololive-scraper-rs/crates/scraper/service/src/maintenance_scheduler.rs
use crate::link_checker::LinkChecker;
use crate::scheduler::format_kst;
use chrono::Utc;
use scraper_core::error::ScraperError;
use scraper_infra::repository::Repository;
use std::time::Duration;
use tokio_util::sync::CancellationToken;
use tracing::{info, warn};

const DEFAULT_EXPIRED_HOUR_KST: u8 = 5;
const DEFAULT_LINK_CHECK_INTERVAL_HOURS: u64 = 12;
const MIN_LINK_CHECK_INTERVAL_HOURS: u64 = 1;
const MAX_LINK_CHECK_INTERVAL_HOURS: u64 = 168; // 1주

#[derive(Debug, Clone)]
pub struct MaintenanceConfig {
    pub expired_hour_kst: u8,
    pub link_check_interval_hours: u64,
}

impl MaintenanceConfig {
    /// link_check_interval_hours를 안전 범위(1~168)로 clamp하여 생성
    pub fn new(expired_hour_kst: u8, link_check_interval_hours: u64) -> Self {
        Self {
            expired_hour_kst,
            link_check_interval_hours: link_check_interval_hours
                .clamp(MIN_LINK_CHECK_INTERVAL_HOURS, MAX_LINK_CHECK_INTERVAL_HOURS),
        }
    }
}

impl Default for MaintenanceConfig {
    fn default() -> Self {
        Self::new(DEFAULT_EXPIRED_HOUR_KST, DEFAULT_LINK_CHECK_INTERVAL_HOURS)
    }
}

#[derive(Clone)]
pub struct MaintenanceScheduler {
    repository: Repository,
    link_checker: LinkChecker,
    config: MaintenanceConfig,
}

impl MaintenanceScheduler {
    pub fn new(
        repository: Repository,
        link_checker: LinkChecker,
        config: MaintenanceConfig,
    ) -> Self {
        Self {
            repository,
            link_checker,
            config,
        }
    }

    pub async fn run(&self, shutdown: CancellationToken) -> Result<(), ScraperError> {
        use crate::scheduler::calculate_next_regular_run_for_hour;

        let link_check_interval = Duration::from_secs(self.config.link_check_interval_hours * 3600);
        let mut link_check_timer = tokio::time::interval(link_check_interval);
        // 첫 tick은 즉시 발생 → 건너뜀 (스타트업 직후 link check 방지)
        link_check_timer.tick().await;

        loop {
            let now = Utc::now();
            let next_expired_run =
                calculate_next_regular_run_for_hour(now, self.config.expired_hour_kst);
            let wait_duration = next_expired_run
                .signed_duration_since(now)
                .to_std()
                .unwrap_or(Duration::ZERO);

            info!(
                next_expired_kst = %format_kst(next_expired_run),
                link_check_interval_hours = self.config.link_check_interval_hours,
                "maintenance scheduler waiting"
            );

            tokio::select! {
                _ = shutdown.cancelled() => {
                    info!("maintenance scheduler received shutdown signal");
                    return Ok(());
                }
                _ = tokio::time::sleep(wait_duration) => {
                    self.run_expired_update().await;
                }
                _ = link_check_timer.tick() => {
                    self.run_link_check().await;
                }
            }
        }
    }

    /// --run-once용: expired update + link check 순차 실행
    pub async fn run_once(&self) -> Result<(), ScraperError> {
        self.run_expired_update().await;
        self.run_link_check().await;
        Ok(())
    }

    async fn run_expired_update(&self) {
        match self.repository.update_expired_events().await {
            Ok(updated) => {
                if updated > 0 {
                    info!(updated_expired = updated, "expired events updated");
                }
            }
            Err(err) => {
                warn!(error = %err, "failed to update expired events");
            }
        }
    }

    async fn run_link_check(&self) {
        match self.link_checker.check_stale_links(&self.repository).await {
            Ok(result) => {
                if result.checked > 0 {
                    info!(
                        checked = result.checked,
                        ok = result.ok,
                        failed = result.failed,
                        blocked = result.blocked,
                        "stale link check completed"
                    );
                }
            }
            Err(err) => {
                warn!(error = %err, "failed to check stale links");
            }
        }
    }
}

File: hololive/hololive-api/internal/planes/llm/internal/service/majorevent/scraper/maintenance_scheduler.go
// Copyright (c) 2025 Kapu
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package scraper

import (
	"context"
	"fmt"
	"log/slog"
	"sync"
	"time"

	"github.com/kapu/hololive-shared/pkg/domain"
	"github.com/kapu/hololive-shared/pkg/panicguard"
)

type maintenanceRepository interface {
	UpdateExpiredEvents(ctx context.Context) (int64, error)
	GetAllActiveEvents(ctx context.Context) ([]*domain.MajorEvent, error)
}

// MaintenanceScheduler는 만료 상태 업데이트와 링크 검증을 주기적으로 수행한다.
type MaintenanceScheduler struct {
	repository  maintenanceRepository
	linkChecker *LinkChecker
	config      MaintenanceConfig
	logger      *slog.Logger
	nowFn       func() time.Time

	stopCh   chan struct{}
	stopOnce sync.Once
	wg       sync.WaitGroup
}

type maintenanceTrigger int

const (
	maintenanceTriggerContextDone maintenanceTrigger = iota
	maintenanceTriggerStop
	maintenanceTriggerExpired
	maintenanceTriggerLinkCheck
)

// NewMaintenanceScheduler는 MaintenanceScheduler를 생성한다.
func NewMaintenanceScheduler(
	repository maintenanceRepository,
	linkChecker *LinkChecker,
	config MaintenanceConfig,
	logger *slog.Logger,
) (*MaintenanceScheduler, error) {
	if repository == nil {
		return nil, fmt.Errorf("new maintenance scheduler: repository is nil")
	}
	if linkChecker == nil {
		return nil, fmt.Errorf("new maintenance scheduler: link checker is nil")
	}
	if logger == nil {
		logger = slog.Default()
	}

	normalized := config
	defaults := DefaultMaintenanceConfig()
	if normalized.RunTimeout <= 0 {
		normalized.RunTimeout = defaults.RunTimeout
	}
	if normalized.LinkCheckInterval <= 0 {
		normalized.LinkCheckInterval = defaults.LinkCheckInterval
	}

	return &MaintenanceScheduler{
		repository:  repository,
		linkChecker: linkChecker,
		config:      normalized,
		logger:      logger,
		nowFn:       time.Now,
		stopCh:      make(chan struct{}),
	}, nil
}

// Start는 유지보수 스케줄러를 시작한다.
func (s *MaintenanceScheduler) Start(ctx context.Context) {
	if s == nil {
		return
	}
	s.wg.Add(1)
	panicguard.Go(s.logger, "major-event-maintenance-scheduler", func() {
		s.run(ctx)
	})
}

// Stop은 유지보수 스케줄러를 종료한다.
func (s *MaintenanceScheduler) Stop() {
	if s == nil {
		return
	}
	s.stopOnce.Do(func() {
		close(s.stopCh)
	})
	s.wg.Wait()
}

func (s *MaintenanceScheduler) run(ctx context.Context) {
	defer s.wg.Done()

	linkTicker := time.NewTicker(s.config.LinkCheckInterval)
	defer linkTicker.Stop()

	for {
		if !s.runMaintenanceCycle(ctx, linkTicker.C) {
			return
		}
	}
}

func (s *MaintenanceScheduler) runMaintenanceCycle(ctx context.Context, linkTicker <-chan time.Time) bool {
	nextExpiredRun := calculateNextRunAtHour(s.now(), s.config.ExpireHourKST)
	waitDuration := max(time.Until(nextExpiredRun), time.Duration(0))

	s.logger.Info(
		"Major event maintenance scheduler waiting",
		slog.String("next_expired_run_kst", formatKST(nextExpiredRun)),
		slog.Duration("wait_duration", waitDuration),
	)

	expiredTimer := time.NewTimer(waitDuration)
	trigger := waitMaintenanceTrigger(ctx, s.stopCh, expiredTimer.C, linkTicker)
	if trigger == maintenanceTriggerContextDone {
		expiredTimer.Stop()
		s.logger.Info("Major event maintenance scheduler stopped by context")
		return false
	}
	if trigger == maintenanceTriggerStop {
		expiredTimer.Stop()
		s.logger.Info("Major event maintenance scheduler stopped")
		return false
	}
	if trigger == maintenanceTriggerExpired {
		s.runExpiredUpdate(ctx)
	}
	if trigger == maintenanceTriggerLinkCheck {
		expiredTimer.Stop()
		s.runLinkCheck(ctx)
	}
	return true
}

func waitMaintenanceTrigger(
	ctx context.Context,
	stopCh <-chan struct{},
	expiredTimer <-chan time.Time,
	linkTicker <-chan time.Time,
) maintenanceTrigger {
	select {
	case <-ctx.Done():
		return maintenanceTriggerContextDone
	case <-stopCh:
		return maintenanceTriggerStop
	case <-expiredTimer:
		return maintenanceTriggerExpired
	case <-linkTicker:
		return maintenanceTriggerLinkCheck
	}
}

func (s *MaintenanceScheduler) runExpiredUpdate(ctx context.Context) {
	runCtx, cancel := context.WithTimeout(ctx, s.config.RunTimeout)
	defer cancel()

	updated, err := s.repository.UpdateExpiredEvents(runCtx)
	if err != nil {
		s.logger.Warn("Major event expired update failed", slog.String("error", err.Error()))
		return
	}
	if updated > 0 {
		s.logger.Info("Major event expired rows updated", slog.Int64("updated_rows", updated))
	}
}

func (s *MaintenanceScheduler) runLinkCheck(ctx context.Context) {
	runCtx, cancel := context.WithTimeout(ctx, s.config.RunTimeout)
	defer cancel()

	events, err := s.repository.GetAllActiveEvents(runCtx)
	if err != nil {
		s.logger.Warn("Major event active event load failed for link check", slog.String("error", err.Error()))
		return
	}
	if len(events) == 0 {
		return
	}

	result, err := s.linkChecker.CheckEvents(runCtx, events)
	if err != nil {
		s.logger.Warn("Major event link check failed", slog.String("error", err.Error()))
		return
	}

	s.logger.Info(
		"Major event link check completed",
		slog.Int("checked", result.Checked),
		slog.Int("ok", result.OK),
		slog.Int("failed", result.Failed),
		slog.Int("blocked", result.Blocked),
	)
}

func (s *MaintenanceScheduler) now() time.Time {
	if s.nowFn != nil {
		return s.nowFn().UTC()
	}
	return time.Now().UTC()
}
